File: control-plane/src/server/mcp-tools/proposeSkillAmp.ts
import { defineTool } from "@github/copilot-sdk";
import { z } from "zod";
import { nanoid } from "nanoid";
import type { StateStore } from "@server/services/stateStore";
import type { SkillAmplification, PolicyRef } from "@shared/types";

export function proposeSkillAmpTool(store: StateStore) {
  return defineTool("propose-skill-amplification", {
    description: "Emit a skill-amplification card for an operator — provides policy context and precedents to support decision-making.",
    parameters: z.object({
      workflowId: z.string().describe("The workflow this amplification card applies to"),
      policyContext: z
        .array(
          z.object({
            title: z.string(),
            snippet: z.string(),
            source: z.string(),
          })
        )
        .optional()
        .describe("Relevant policy references to surface to the operator"),
      precedents: z
        .array(
          z.object({
            workflowId: z.string(),
            outcome: z.string(),
            rationale: z.string(),
          })
        )
        .optional()
        .describe("Past similar decisions and their rationale"),
      recommendedApproach: z
        .string()
        .describe("Concise recommended approach for the operator"),
    }),
    skipPermission: true,
    handler: async ({ workflowId, policyContext, precedents, recommendedApproach }) => {
      const id = `AMP-${nanoid(8)}`;
      const amp: SkillAmplification = {
        id,
        workflowId,
        policyContext: (policyContext ?? []) as PolicyRef[],
        precedents: precedents ?? [],
        recommendedApproach,
        createdAt: Date.now(),
      };
      store.appendAmplification(workflowId, amp);
      return { amplificationId: id };
    },
  });
}

File: control-plane-py/src/server/mcp_tools/propose_skill_amp.py
# src/server/mcp_tools/propose_skill_amp.py
from __future__ import annotations
import json
import time
from typing import Any
from pydantic import BaseModel
from nanoid import generate as nanoid
from copilot.tools import define_tool, ToolInvocation, ToolResult
from src.server.services.state_store import StateStore
from src.shared.types import SkillAmplification, PolicyRef


class ProposeSkillAmpParams(BaseModel):
    workflow_id: str
    policy_context: list[dict[str, Any]] | None = None
    precedents: list[dict[str, Any]] | None = None
    recommended_approach: str


def make_propose_skill_amp_tool(store: StateStore):
    @define_tool(description="Emit a coach card for an operator with policy context and precedents.", skip_permission=True)
    def propose_skill_amplification(params: ProposeSkillAmpParams, invocation: ToolInvocation) -> ToolResult:
        a = SkillAmplification(
            id=f"AMP-{nanoid(size=8)}",
            workflow_id=params.workflow_id,
            policy_context=[PolicyRef(**p) for p in (params.policy_context or [])],
            precedents=params.precedents or [],
            recommended_approach=params.recommended_approach,
            created_at=time.time(),
        )
        store.append_amplification(params.workflow_id, a)
        return ToolResult(text_result_for_llm=json.dumps({"amplification_id": a.id}), result_type="success")

    return propose_skill_amplification
